_benchmark/lib/workspace.dart

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2024, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:convert';
+import 'dart:io';
+
+import 'config.dart';
+
+/// A temporary workspace for benchmarking `build_runner`.
+class Workspace {
+  final Config config;
+  final String name;
+
+  Directory get directory =>
+      Directory.fromUri(config.rootDirectory.uri.resolve(name));
+
+  Workspace({required this.config, required this.name, bool clean = true}) {
+    if (clean) {
+      if (directory.existsSync()) directory.deleteSync(recursive: true);
+      directory.createSync(recursive: true);
+    }
+  }
+
+  /// Writes [source] to [path].
+  void write(String path, {required String source}) {
+    final file = File.fromUri(directory.uri.resolve(path));
+    file.parent.createSync(recursive: true);
+    file.writeAsStringSync(source);
+  }
+
+  /// Edits the file at [path] to replace `CACHEBUSTER` with a timestamp.
+  ///
+  /// Throws `StateError` if the string is not present to replace.
+  void edit(String path) {
+    final file = File.fromUri(directory.uri.resolve(path));
+    final source = file.readAsStringSync();
+    final edited = source.replaceAll(
+      'CACHEBUSTER',
+      'CACHEBUSTER${DateTime.now().microsecondsSinceEpoch.toString()}',
+    );
+    if (source == edited) {
+      throw StateError('Edit of "$path" failed: no CACHEBUSTER in source.');
+    }
+    file.writeAsStringSync(edited);
+  }
+
+  /// Builds and measures performance.
+  PendingResult measure() {
+    final result = PendingResult();
+    _measure(result);
+    return result;
+  }
+
+  Future<void> _measure(PendingResult result) async {
+    // Clean build.
+    final stopwatch = Stopwatch()..start();
+    var process = await Process.start('dart', [
+      'run',
+      'build_runner',
+      'build',
+      '-d',
+    ], workingDirectory: directory.path);
+    var exitCode = await process.exitCode;
+    result.cleanBuildTime = stopwatch.elapsed;
+    if (exitCode != 0) {
+      final stdout = await process.stdout.transform(utf8.decoder).join();
+      final stderr = await process.stderr.transform(utf8.decoder).join();
+      result.failure = 'Initial build failed:\n$stdout\n$stderr';
+      return;
+    }
+
+    // Build with no changes.
+    stopwatch.reset();
+    process = await Process.start('dart', [
+      'run',
+      'build_runner',
+      'build',
+      '-d',
+    ], workingDirectory: directory.path);
+    exitCode = await process.exitCode;
+    result.noChangesBuildTime = stopwatch.elapsed;
+    if (exitCode != 0) {
+      final stdout = await process.stdout.transform(utf8.decoder).join();
+      final stderr = await process.stderr.transform(utf8.decoder).join();
+      result.failure = 'No changes build failed:\n$stdout\n$stderr';
+      return;
+    }
+
+    // Incremental build after a change.
```

Comment:
Fwiw, this kind of incremental build is probably a lot slower than a "watch" mode incremental build, because this is going to do a lot of startup work (reading asset graph, scanning and hashing all files to check for changes, etc).

Ideally, we would be able to exclude that (or separate it out).

Reply:
Yup yup, we probably want finer-grained benchmarks as we go.

For now this seems to give a clear enough signal :)

Thanks.